src/transport/manager/mod.rs

```diff
@@ -648,77 +567,44 @@ impl TransportManager {
                     "invalid protocol"
                 );
 
-                return Err(Error::TransportNotSupported(record.address().clone()));
+                return Err(Error::TransportNotSupported(
+                    address_record.address().clone(),
+                ));
             }
         };
 
         // when constructing `AddressRecord`, `PeerId` was verified to be part of the address
         let remote_peer_id =
-            PeerId::try_from_multiaddr(record.address()).expect("`PeerId` to exist");
+            PeerId::try_from_multiaddr(address_record.address()).expect("`PeerId` to exist");
 
         // set connection id for the address record and put peer into `Dialing` state
         let connection_id = self.next_connection_id();
-        record.set_connection_id(connection_id);
+        let dial_record = ConnectionRecord {
+            address: address_record.address().clone(),
+            connection_id,
+        };
 
         {
             let mut peers = self.peers.write();
 
-            match peers.entry(remote_peer_id) {
-                Entry::Occupied(occupied) => {
-                    let context = occupied.into_mut();
+            let context = peers.entry(remote_peer_id).or_insert_with(|| PeerContext::default());
 
-                    context.addresses.insert(record.clone());
+            // Keep the provided record around for possible future dials.
+            context.addresses.insert(address_record.clone());
```

Comment:
Won't this lead to duplicate records if we repeatedly dial an address?

Reply:
With the refactored address store we can update address scores without duplicates 🙏